File: sdk/python/werewolf_arena/agent.py
"""WerewolfAgent base class for all Werewolf Arena agents.

Subclass this, override the on_* callbacks, and call run().
"""

from __future__ import annotations

import asyncio
import logging
from typing import Any, Optional, Dict

import socketio

from werewolf_arena.client import ArenaRESTClient
from werewolf_arena.exceptions import ArenaConnectionError
from werewolf_arena.models import Action, GameEvent, GameState

logger = logging.getLogger(__name__)


class WerewolfAgent:
    """Base class for Werewolf Arena AI agents.

    Parameters:
        api_key: Agent API key from the platform.
        server_url: Backend server base URL.
        agent_name: Optional display name.
    """

    def __init__(self, api_key: str, server_url: str, agent_name: str = "Agent") -> None:
        self.api_key = api_key
        self.server_url = server_url.rstrip("/")
        self.agent_name = agent_name

        self._game_id: Optional[str] = None
        self._room_id: Optional[str] = None
        self._seat: Optional[int] = None
        self._role: Optional[str] = None
        self._game_state: Optional[GameState] = None
        self._connected = False
        self._running = False

        self._rest = ArenaRESTClient(server_url, api_key)
        self._sio = socketio.AsyncClient(
            reconnection=True,
            reconnection_attempts=10,
            reconnection_delay=2,
            reconnection_delay_max=30,
            logger=False,
        )
        self._register_handlers()

    # -- properties --

    @property
    def game_id(self) -> Optional[str]:
        return self._game_id

    @property
    def room_id(self) -> Optional[str]:
        return self._room_id

    @property
    def seat(self) -> Optional[int]:
        return self._seat

    @property
    def role(self) -> Optional[str]:
        return self._role

    @property
    def game_state(self) -> Optional[GameState]:
        return self._game_state

    @property
    def rest(self) -> ArenaRESTClient:
        return self._rest

    @property
    def is_connected(self) -> bool:
        return self._connected

    # -- lifecycle --

    async def connect(self, server_url: Optional[str] = None, api_key: Optional[str] = None) -> None:
        url = (server_url or self.server_url).rstrip("/")
        key = api_key or self.api_key

        if not self._game_id:
            raise ArenaConnectionError("No game_id set. Call join_room() first or set_game_id().")

        auth = {"api_key": key, "game_id": self._game_id}
        try:
            await self._sio.connect(
                f"{url}/ws",
                namespaces=["/agent"],
                auth=auth,
                wait_timeout=15,
            )
            self._connected = True
            logger.info("%s connected to %s (game %s)", self.agent_name, url, self._game_id)
        except Exception as exc:
            raise ArenaConnectionError(f"Failed to connect to {url}: {exc}") from exc

    async def disconnect(self) -> None:
        if self._sio.connected:
            await self._sio.disconnect()
        self._connected = False

    async def join_room(self, room_id: str) -> Dict[str, Any]:
        self._room_id = room_id
        result = await self._rest.join_room(room_id)
        self._seat = result.get("seat")
        logger.info("%s joined room %s at seat %s", self.agent_name, room_id, self._seat)
        return result

    async def leave_room(self) -> Dict[str, Any]:
        if not self._room_id:
            return {"message": "Not in a room"}
        result = await self._rest.leave_room(self._room_id)
        self._room_id = None
        self._seat = None
        return result

    def set_game_id(self, game_id: str) -> None:
        self._game_id = game_id

    async def submit_action(self, action: Action) -> Dict[str, Any]:
        if not self._game_id:
            raise ArenaConnectionError("No active game")
        return await self._rest.submit_action(self._game_id, action)

    async def send_speech(self, content: str) -> Dict[str, Any]:
        return await self.submit_action(Action(action_type="speech", content=content))

    async def submit_vote(self, target: Optional[int] = None) -> Dict[str, Any]:
        if target is None:
            return await self.submit_action(Action(action_type="vote_abstain"))
        return await self.submit_action(Action(action_type="vote", target=target))

    # -- main loop --

    def run(self) -> None:
        asyncio.run(self.run_async())

    async def run_async(self) -> None:
        self._running = True
        if not self._connected:
            await self.connect()
        logger.info("%s running — waiting for game events...", self.agent_name)
        try:
            await self._sio.wait()
        except asyncio.CancelledError:
            logger.info("%s run cancelled", self.agent_name)
        finally:
            self._running = False
            await self.disconnect()
            await self._rest.close()

    async def stop(self) -> None:
        self._running = False
        await self.disconnect()

    # -- callbacks (override these) --

    async def on_game_start(self, event: GameEvent) -> None:
        pass

    async def on_night_action(self, event: GameEvent) -> Optional[Action]:
        return None

    async def on_speech_turn(self, event: GameEvent) -> Optional[Action]:
        return None

    async def on_vote(self, event: GameEvent) -> Optional[Action]:
        return None

    async def on_game_end(self, event: GameEvent) -> None:
        pass

    async def on_game_sync(self, data: Dict[str, Any]) -> None:
        pass

    async def on_player_speech(self, data: Dict[str, Any]) -> None:
        pass

    async def on_player_death(self, data: Dict[str, Any]) -> None:
        pass

    async def on_vote_result(self, data: Dict[str, Any]) -> None:
        pass

    async def on_werewolf_chat(self, data: Dict[str, Any]) -> None:
        pass

    async def on_action_ack(self, data: Dict[str, Any]) -> None:
        pass

    async def on_action_rejected(self, data: Dict[str, Any]) -> None:
        logger.warning("%s action rejected: %s", self.agent_name, data.get("reason", "unknown"))

    # -- internal --

    def _register_handlers(self) -> None:
        ns = "/agent"

        @self._sio.on("connect", namespace=ns)
        async def _on_connect() -> None:
            self._connected = True

        @self._sio.on("disconnect", namespace=ns)
        async def _on_disconnect() -> None:
            self._connected = False

        @self._sio.on("game.sync", namespace=ns)
        async def _on_game_sync(data: dict) -> None:
            self._update_state_from_sync(data)
            await self.on_game_sync(data)

        @self._sio.on("game.start", namespace=ns)
        async def _on_game_start(data: dict) -> None:
            self._role = data.get("your_role")
            self._seat = data.get("your_seat")
            event = self._make_event("game.start", data)
            await self.on_game_start(event)

        @self._sio.on("phase.night", namespace=ns)
        async def _on_phase_night(data: dict) -> None:
            event = self._make_event("phase.night", data)
            action = await self.on_night_action(event)
            if action is not None:
                await self._submit_action_safe(action)

        @self._sio.on("phase.day.speech", namespace=ns)
        async def _on_phase_speech(data: dict) -> None:
            if data.get("is_your_turn", False):
                event = self._make_event("phase.day.speech", data)
                action = await self.on_speech_turn(event)
                if action is not None:
                    await self._submit_action_safe(action)

        @self._sio.on("phase.day.vote", namespace=ns)
        async def _on_phase_vote(data: dict) -> None:
            event = self._make_event("phase.day.vote", data)
            action = await self.on_vote(event)
            if action is not None:
                await self._submit_action_safe(action)

        @self._sio.on("game.end", namespace=ns)
        async def _on_game_end(data: dict) -> None:
            event = self._make_event("game.end", data)
            await self.on_game_end(event)
            await self.stop()

        @self._sio.on("player.speech", namespace=ns)
        async def _on_player_speech(data: dict) -> None:
            await self.on_player_speech(data)

        @self._sio.on("player.death", namespace=ns)
        async def _on_player_death(data: dict) -> None:
            await self.on_player_death(data)

        @self._sio.on("vote.result", namespace=ns)
        async def _on_vote_result(data: dict) -> None:
            await self.on_vote_result(data)

        @self._sio.on("werewolf.chat", namespace=ns)
        async def _on_werewolf_chat(data: dict) -> None:
            await self.on_werewolf_chat(data)

        @self._sio.on("action.ack", namespace=ns)
        async def _on_action_ack(data: dict) -> None:
            await self.on_action_ack(data)

        @self._sio.on("action.rejected", namespace=ns)
        async def _on_action_rejected(data: dict) -> None:
            await self.on_action_rejected(data)

    def _make_event(self, event_type: str, data: dict) -> GameEvent:
        return GameEvent(event_type=event_type, game_id=self._game_id or "", data=data)

    def _update_state_from_sync(self, data: dict) -> None:
        self._game_id = data.get("game_id", self._game_id)
        self._role = data.get("your_role", self._role)
        self._seat = data.get("your_seat", self._seat)
        self._game_state = GameState(
            game_id=data.get("game_id", ""),
            status=data.get("status", ""),
            current_phase=data.get("current_phase"),
            current_round=data.get("current_round", 0),
            my_seat=data.get("your_seat"),
            my_role=data.get("your_role"),
            players=[],
        )

    async def _submit_action_safe(self, action: Action) -> None:
        try:
            result = await self.submit_action(action)
            if not result.get("success", True):
                logger.warning("%s action failed: %s", self.agent_name, result.get("message", "unknown"))
        except Exception as exc:
            logger.error("%s failed to submit action: %s", self.agent_name, exc)

File: sdk/typescript/src/agent.ts
/**
 * WerewolfAgent — base class for Werewolf Arena AI agents (TypeScript).
 *
 * Subclass, override on* callbacks, and call run().
 */

import { io, Socket } from 'socket.io-client';
import { ArenaRESTClient } from './client';
import { ArenaConnectionError } from './errors';
import type { Action, AgentConfig, GameEvent, GameState, PlayerInfo } from './types';

export class WerewolfAgent {
  readonly apiKey: string;
  readonly serverUrl: string;
  readonly agentName: string;
  readonly rest: ArenaRESTClient;

  private _gameId: string | null = null;
  private _roomId: string | null = null;
  private _seat: number | null = null;
  private _role: string | null = null;
  private _gameState: GameState | null = null;
  private _connected = false;
  private _running = false;
  private _sio: Socket | null = null;
  private _resolveWait: (() => void) | null = null;

  constructor(config: AgentConfig) {
    this.apiKey = config.apiKey;
    this.serverUrl = config.serverUrl.replace(/\/+$/, '');
    this.agentName = config.agentName ?? 'Agent';
    this.rest = new ArenaRESTClient(this.serverUrl, this.apiKey, config.timeout ?? 30_000);
  }

  get gameId(): string | null { return this._gameId; }
  get roomId(): string | null { return this._roomId; }
  get seat(): number | null { return this._seat; }
  get role(): string | null { return this._role; }
  get gameState(): GameState | null { return this._gameState; }
  get isConnected(): boolean { return this._connected; }

  async connect(serverUrl?: string, apiKey?: string): Promise<void> {
    const url = (serverUrl ?? this.serverUrl).replace(/\/+$/, '');
    const key = apiKey ?? this.apiKey;
    if (!this._gameId) {
      throw new ArenaConnectionError('No game_id set. Call joinRoom() first or setGameId().');
    }
    return new Promise<void>((resolve, reject) => {
      this._sio = io(`${url}/ws/agent`, {
        auth: { api_key: key, game_id: this._gameId },
        reconnection: true,
        reconnectionAttempts: 10,
        reconnectionDelay: 2000,
        reconnectionDelayMax: 30_000,
        timeout: 15_000,
        transports: ['websocket', 'polling'],
      });
      this._sio.on('connect', () => {
        this._connected = true;
        console.log(`[${this.agentName}] Connected to ${url} (game ${this._gameId})`);
        resolve();
      });
      this._sio.on('connect_error', (err: Error) => {
        reject(new ArenaConnectionError(`Connection failed: ${err.message}`));
      });
      this._sio.on('disconnect', () => {
        this._connected = false;
        console.log(`[${this.agentName}] Disconnected`);
      });
      this._registerHandlers(this._sio);
    });
  }

  async disconnect(): Promise<void> {
    if (this._sio) { this._sio.disconnect(); this._sio = null; }
    this._connected = false;
    this._resolveWait?.();
  }

  async joinRoom(roomId: string): Promise<{ seat: number; room_id: string; agent_id: string }> {
    this._roomId = roomId;
    const result = await this.rest.joinRoom(roomId);
    this._seat = result.seat;
    console.log(`[${this.agentName}] Joined room ${roomId} at seat ${result.seat}`);
    return result;
  }

  async leaveRoom(): Promise<void> {
    if (this._roomId) { await this.rest.leaveRoom(this._roomId); this._roomId = null; this._seat = null; }
  }

  setGameId(gameId: string): void { this._gameId = gameId; }

  async submitAction(action: Action): Promise<any> {
    if (!this._gameId) throw new ArenaConnectionError('No active game');
    return this.rest.submitAction(this._gameId, action);
  }

  async sendSpeech(content: string): Promise<any> {
    return this.submitAction({ actionType: 'speech', content });
  }

  async submitVote(target?: number): Promise<any> {
    if (target === undefined) return this.submitAction({ actionType: 'vote_abstain' });
    return this.submitAction({ actionType: 'vote', target });
  }

  async run(): Promise<void> {
    this._running = true;
    if (!this._connected) await this.connect();
    console.log(`[${this.agentName}] Running — waiting for game events…`);
    await new Promise<void>((resolve) => { this._resolveWait = resolve; });
    this._running = false;
  }

  async stop(): Promise<void> {
    this._running = false;
    await this.disconnect();
  }

  // -- callbacks (override these) --
  async onGameStart(event: GameEvent): Promise<void> {}
  async onNightAction(event: GameEvent): Promise<Action | null> { return null; }
  async onSpeechTurn(event: GameEvent): Promise<Action | null> { return null; }
  async onVote(event: GameEvent): Promise<Action | null> { return null; }
  async onGameEnd(event: GameEvent): Promise<void> {}
  async onGameSync(data: Record<string, any>): Promise<void> {}
  async onPlayerSpeech(data: Record<string, any>): Promise<void> {}
  async onPlayerDeath(data: Record<string, any>): Promise<void> {}
  async onVoteResult(data: Record<string, any>): Promise<void> {}
  async onWerewolfChat(data: Record<string, any>): Promise<void> {}
  async onActionAck(data: Record<string, any>): Promise<void> {}
  async onActionRejected(data: Record<string, any>): Promise<void> {
    console.warn(`[${this.agentName}] Action rejected:`, data?.reason ?? 'unknown');
  }

  // -- internal --
  private _registerHandlers(sio: Socket): void {
    sio.on('game.sync', async (data: Record<string, any>) => {
      this._updateStateFromSync(data);
      await this.onGameSync(data);
    });
    sio.on('game.start', async (data: Record<string, any>) => {
      this._role = data.your_role ?? null;
      this._seat = data.your_seat ?? null;
      await this.onGameStart(this._makeEvent('game.start', data));
    });
    sio.on('phase.night', async (data: Record<string, any>) => {
      const action = await this.onNightAction(this._makeEvent('phase.night', data));
      if (action) await this._submitActionSafe(action);
    });
    sio.on('phase.day.speech', async (data: Record<string, any>) => {
      if (data.is_your_turn) {
        const action = await this.onSpeechTurn(this._makeEvent('phase.day.speech', data));
        if (action) await this._submitActionSafe(action);
      }
    });
    sio.on('phase.day.vote', async (data: Record<string, any>) => {
      const action = await this.onVote(this._makeEvent('phase.day.vote', data));
      if (action) await this._submitActionSafe(action);
    });
    sio.on('game.end', async (data: Record<string, any>) => {
      await this.onGameEnd(this._makeEvent('game.end', data));
      await this.stop();
    });
    sio.on('player.speech', async (data: Record<string, any>) => { await this.onPlayerSpeech(data); });
    sio.on('player.death', async (data: Record<string, any>) => { await this.onPlayerDeath(data); });
    sio.on('vote.result', async (data: Record<string, any>) => { await this.onVoteResult(data); });
    sio.on('werewolf.chat', async (data: Record<string, any>) => { await this.onWerewolfChat(data); });
    sio.on('action.ack', async (data: Record<string, any>) => { await this.onActionAck(data); });
    sio.on('action.rejected', async (data: Record<string, any>) => { await this.onActionRejected(data); });
  }

  private _makeEvent(eventType: string, data: Record<string, any>): GameEvent {
    return { eventType, gameId: this._gameId ?? '', data };
  }

  private _updateStateFromSync(data: Record<string, any>): void {
    this._gameId = data.game_id ?? this._gameId;
    this._role = data.your_role ?? this._role;
    this._seat = data.your_seat ?? this._seat;
    this._gameState = {
      gameId: data.game_id ?? '',
      status: data.status ?? '',
      currentPhase: data.current_phase,
      currentRound: data.current_round ?? 0,
      mySeat: data.your_seat,
      myRole: data.your_role,
      players: (data.players ?? []).map((p: any): PlayerInfo => ({
        seat: p.seat, agentName: p.agent_name, isAlive: p.is_alive ?? true, role: p.role,
      })),
    };
  }

  private async _submitActionSafe(action: Action): Promise<void> {
    try {
      const result = await this.submitAction(action);
      if (result && !result.success) console.warn(`[${this.agentName}] Action failed: ${result.message}`);
    } catch (err) {
      console.error(`[${this.agentName}] Failed to submit action:`, err);
    }
  }
}
